File: 1_못푼문제/프로그래머스Lv3-표 편집.py
import copy
def solution(n, k, cmd):
    table = [i for i in range(n)]
    snapshots = [copy.copy(table)]

    for x in cmd:
        commands = x.split(" ")
        command = commands[0]
        if command == "U" or command == "D":
            shift = int(commands[1])
            if command == "U":
                k -= shift
                if k <= 0:
                    k = 0
            else:
                k += shift
                if k >= n - 1:
                    k = n - 1

        elif command == "C":
            snapshots.append(copy.copy(table))
            del table[k]
            last_idx = len(table) - 1
            if k > last_idx:
                k = last_idx

        elif command == "Z":
            snapshot = snapshots.pop()
            if table[k] != snapshot[k]:
                k += 1
            table = snapshot
        
    result = ""

    for i in range(n):
        if i in table:
            result += "O"
        else:
            result += "X"

    return result

"""
시간초과가 나는 이유는 아마 del 사용(O(N)이 추가), 마지막에 결과를 구하는 과정에서 O(N^2)이 발생했기 때문같다.
정확성 테스트
테스트 1 〉 통과 (0.04ms, 10.4MB)
테스트 2 〉 통과 (0.03ms, 10.4MB)
테스트 3 〉 통과 (0.04ms, 10.4MB)
테스트 4 〉 통과 (0.03ms, 10.4MB)
테스트 5 〉 통과 (0.12ms, 10.3MB)
테스트 6 〉 통과 (0.13ms, 10.5MB)
테스트 7 〉 통과 (0.12ms, 10.4MB)
테스트 8 〉 통과 (0.15ms, 10.4MB)
테스트 9 〉 통과 (0.13ms, 10.4MB)
테스트 10 〉 통과 (0.12ms, 10.4MB)
테스트 11 〉 통과 (1.60ms, 10.4MB)
테스트 12 〉 통과 (1.56ms, 10.6MB)
테스트 13 〉 통과 (1.57ms, 10.5MB)
테스트 14 〉 통과 (1.58ms, 10.4MB)
테스트 15 〉 통과 (1.62ms, 10.6MB)
테스트 16 〉 통과 (1.60ms, 10.7MB)
테스트 17 〉 통과 (5.63ms, 10.6MB)
테스트 18 〉 통과 (5.79ms, 10.5MB)
테스트 19 〉 통과 (5.83ms, 10.4MB)
테스트 20 〉 통과 (6.32ms, 10.7MB)
테스트 21 〉 통과 (6.33ms, 10.9MB)
테스트 22 〉 통과 (6.80ms, 11MB)
테스트 23 〉 통과 (0.03ms, 10.4MB)
테스트 24 〉 통과 (0.03ms, 10.6MB)
테스트 25 〉 통과 (0.03ms, 10.4MB)
테스트 26 〉 통과 (0.03ms, 10.4MB)
테스트 27 〉 통과 (0.04ms, 10.4MB)
테스트 28 〉 통과 (0.04ms, 10.4MB)
테스트 29 〉 통과 (0.06ms, 10.4MB)
테스트 30 〉 통과 (0.05ms, 10.4MB)
효율성 테스트
테스트 1 〉 실패 (시간 초과)
테스트 2 〉 실패 (시간 초과)
테스트 3 〉 실패 (시간 초과)
테스트 4 〉 실패 (시간 초과)
테스트 5 〉 실패 (시간 초과)
테스트 6 〉 실패 (시간 초과)
테스트 7 〉 실패 (시간 초과)
테스트 8 〉 실패 (시간 초과)
테스트 9 〉 실패 (시간 초과)
테스트 10 〉 실패 (시간 초과)
"""

# 링크드리스트 풀이

File: 1_못푼문제/프로그래머스Lv3-표 편집.js
class Node {
  constructor(value) {
    this.value = value
    this.prev = this.next = null
  }
}

class DoubleLinkedList {
  constructor() {
    this.head = this.tail = null
    this.size = 0
  }

  clear() {
    this.head = this.tail = null
    this.size = 0
  }

  desc(inverse = false) {
    const nodes = []
    let currentNode = this.head
    while (currentNode !== null) {
      nodes.push(currentNode.value)
      currentNode = currentNode.next
    }
    return inverse ? nodes.reverse() : nodes
  }

  add(value) {
    const newNode = new Node(value)

    if (this.head === null) {
      this.head = newNode
      this.tail = newNode
    } else {
      this.tail.next = newNode
      newNode.prev = this.tail
      this.tail = newNode
    }

    this.size++
  }

  insert(position, value) {
    if (typeof position !== 'number' || position < 0 || position > this.size) {
      return
    }

    let newNode = new Node(value)

    if (position === 0) {
      if (this.head === null) {
        this.head = newNode
        this.tail = newNode
      } else {
        newNode.next = this.head
        this.head.prev = newNode
        this.head = newNode
      }
    } else if (position === this.size) {
      this.tail.next = newNode
      newNode.prev = this.tail
      this.tail = newNode
    } else {
      let index = 0
      let currentNode = this.head
      let prevNode
      while (index++ < position) {
        prevNode = currentNode
        currentNode = currentNode.next
      }
      newNode.next = currentNode
      prevNode.next = newNode
      currentNode.prev = newNode
      newNode.prev = prevNode
    }
    this.size++
  }

  remove(value) {
    let currentNode = this.head
    let prevNode = currentNode

    while (currentNode.value !== value && currentNode.next !== null) {
      prevNode = currentNode
      currentNode = currentNode.next
    }

    if (currentNode.value !== value) {
      return undefined
    }

    if (currentNode === this.head) {
      this.head = currentNode.next
      if (this.size === 1) {
        this.tail = null
      } else {
        this.head.prev = null
      }
    } else if (currentNode === this.tail) {
      this.tail = currentNode.prev
      this.tail.next = null
    } else {
      prevNode.next = currentNode.next
      currentNode.next.prev = prevNode
    }
    this.size--

    return currentNode
  }

  removeAt(position) {
    if (typeof position !== 'number' || position < 0 || position >= this.size) {
      return
    }

    let currentNode = this.head
    let index = 0
    let prevNode

    if (position === 0) {
      this.head = currentNode.next
      if (this.size === 1) {
        this.tail = null
      } else {
        this.head.prev = null
      }
    } else if (position === this.size - 1) {
      currentNode = this.tail
      this.tail = currentNode.prev
      this.tail.next = null
    } else {
      while (index++ < position) {
        prevNode = currentNode
        currentNode = currentNode.next
      }
      prevNode.next = currentNode.next
      currentNode.next.prev = prevNode
    }

    this.size--
    return currentNode
  }

  indexOf(value) {
    let currentNode = this.head
    let index = 0

    while (currentNode !== null) {
      if (currentNode.value === value) {
        return index
      }
      index++
      currentNode = currentNode.next
    }
    return -1
  }

  remove2(value) {
    const index = this.indexOf(value)
    return index > -1 ? this.removeAt(index) : undefined
  }
}

function solution(n, k, cmd) {
  const DL = new DoubleLinkedList()
  const snapshots = []
  for (let i = 0; i < n; i++) {
    DL.add(i)
  }

  for (const c of cmd) {
    const commands = c.split(' ')
    const command = commands[0]
    if (command === 'U' || command === 'D') {
      const shift = parseInt(commands[1], 10)
      if (command === 'U') {
        k - shift < 0 ? (k = 0) : (k -= shift)
      } else {
        k + shift > DL.size - 1 ? (k = DL.size - 1) : (k += shift)
      }
    } else if (command === 'C') {
      const removedNode = DL.removeAt(k)
      snapshots.push([k, removedNode.prev.value, removedNode.value])
      if (k > DL.size - 1) {
        k = DL.size - 1
      }
    } else {
      const [prevK, prevValue, removedValue] = snapshots.pop()
      DL.insert(prevK, removedValue)
      if (k > prevK) {
        k++
      }
    }
  }

  const result = Array.from({ length: n }, () => 'X')

  for (const v of DL.desc()) {
    result[v] = 'O'
  }

  return result.join('')
}
// 정확성  테스트
// 테스트 1 〉	통과 (0.53ms, 30.1MB)
// 테스트 2 〉	실패 (런타임 에러)
// 테스트 3 〉	통과 (0.53ms, 30MB)
// 테스트 4 〉	통과 (0.55ms, 30.1MB)
// 테스트 5 〉	실패 (0.59ms, 30.1MB)
// 테스트 6 〉	통과 (0.63ms, 30.1MB)
// 테스트 7 〉	실패 (0.65ms, 30MB)
// 테스트 8 〉	실패 (0.61ms, 30MB)
// 테스트 9 〉	실패 (0.67ms, 30.1MB)
// 테스트 10 〉	실패 (런타임 에러)
// 테스트 11 〉	실패 (2.62ms, 32.5MB)
// 테스트 12 〉	실패 (2.87ms, 32.1MB)
// 테스트 13 〉	실패 (2.98ms, 32.3MB)
// 테스트 14 〉	실패 (런타임 에러)
// 테스트 15 〉	실패 (런타임 에러)
// 테스트 16 〉	실패 (런타임 에러)
// 테스트 17 〉	실패 (런타임 에러)
// 테스트 18 〉	실패 (런타임 에러)
// 테스트 19 〉	실패 (런타임 에러)
// 테스트 20 〉	실패 (런타임 에러)
// 테스트 21 〉	실패 (런타임 에러)
// 테스트 22 〉	실패 (런타임 에러)
// 테스트 23 〉	실패 (0.51ms, 30.2MB)
// 테스트 24 〉	실패 (0.52ms, 30MB)
// 테스트 25 〉	통과 (0.50ms, 29.9MB)
// 테스트 26 〉	실패 (런타임 에러)
// 테스트 27 〉	실패 (런타임 에러)
// 테스트 28 〉	실패 (런타임 에러)
// 테스트 29 〉	통과 (0.54ms, 29.9MB)
// 테스트 30 〉	실패 (0.53ms, 30MB)
// 효율성  테스트
// 테스트 1 〉	실패 (시간 초과)
// 테스트 2 〉	실패 (시간 초과)
// 테스트 3 〉	실패 (시간 초과)
// 테스트 4 〉	실패 (시간 초과)
// 테스트 5 〉	실패 (시간 초과)
// 테스트 6 〉	실패 (시간 초과)
// 테스트 7 〉	실패 (시간 초과)
// 테스트 8 〉	실패 (런타임 에러)
// 테스트 9 〉	실패 (시간 초과)
// 테스트 10 〉	실패 (시간 초과)
